tensorly/decomposition/mda.py

```diff
@@ -0,0 +1,165 @@
+    factors : list
+        list of the learnt projection matrices for each mode
+
+    Notes
+    -----
+
+    This implementation computes the *Constrained Multilinear Discriminant Analysis* (CMDA) solution as presented in [1].
+
+    Given the learnt factor matrices, one can then compute the projection along all modes with::
+
+      factors = mda(X_train, y_train, ranks, n)
+      Z = tl.tenalg.multi_mode_dot(X_train, factors, modes=[1, 2], transpose=True)
+
+    - [1] Q. Li and D. Schonfeld, "Multilinear Discriminant Analysis for Higher-Order Tensor Data Classification," in IEEE Transactions on Pattern Analysis and Machine Intelligence, vol. 36, no. 12, pp. 2524-2537, 1 Dec. 2014, doi: 10.1109/TPAMI.2014.2342214.
+
+    """
+
+    ###############
+    # check correct # of ranks have been supplied
+    ###############
+    assert len(ranks) == len(T.shape(X)[1:]), 'Expected number of ranks: {}. \
+        But number supplied is {}'.format(len(T.shape(X)[1:]), len(ranks))
+
+    backend = T.get_backend()
+
+    global_mean = T.mean(X, axis=0)
+    class_means = []
+
+    # ith element will contain a list of the indices of training data with class label i
+    class_idx = [[] for _ in range(len(set(y)))]
+
+    # store the training data's class label at class index
+    for i, label in enumerate(y):
+        class_idx[label] += [i]
+
+    # store the mean of all tensor samples with label i
+    for i in range(len(set(y))):
+        # tensorflow is only backend to not support indexing into tensor with a list
+        if backend == 'tensorflow':
+            class_means += [T.mean(T.tensor([X[j] for j in class_idx[i]]), axis=0)]
+        else:
+            class_means += [T.mean(X[class_idx[i], ...], axis=0)]
+
+    # the first mode is the 'sample' mode
+    num_modes = len(T.shape(X)) - 1
+
+    # initialise the factor matrices as 1-matrices
+    factors = [T.ones((dim, T.shape(X)[i + 1]), **T.context(X))
+               for i, dim in enumerate(list(T.shape(X))[1:])]
+
+    for t in range(1, n_iters + 1):
+        # for each iteration compute partial projections for mode k,
+        # i.e. project along all modes but k.
+        for k in range(num_modes):
+            B_scat, W_scat = compute_modek_wb_scatters(X, k, factors, global_mean, class_means, class_idx)
+
+            # first compute the inverse of the scatter matrix
+            # i.e. solve SX=I for X
+            W_scat_inv = T.solve(W_scat, T.eye(T.shape(W_scat)[0]))
+
+            ###################################################
+            # *Constrained Multilinear Discriminant Analysis* (CMDA) [1] solution for factor matrix U_k is given by
+            # top `rank' number of left-singular vectors of W^{_1}B.
+            # --
+            # [1] Q. Li et al. "Multilinear Discriminant Analysis for Higher-Order Tensor Data Classification"
+            ###################################################
+            U, _, _ = T.partial_svd(T.dot(W_scat_inv,  B_scat))
```

Comment:
We should probably be using SVD_FUNS but this is being refactored in #217 and should be easier to use when that gets merged.